lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -524,6 +524,19 @@ ChangeResult TypeAnalyzer::visitOperation(
     return incorporateKnowledge(op->getResult(0), knowledge);
   }
 
+  // Dtype is always float32, except for float64.
+  if (isa<AtenTanhOp, AtenExpOp, AtenSinOp, AtenCosOp,
+          AtenSigmoidOp, AtenReciprocalOp, AtenLogOp, AtenSqrtOp,
+          AtenLog2Op, AtenRsqrtOp, AtenErfOp>(op)) {
+    ValueKnowledge knowledge =
+        ValueKnowledge::getNotNonePessimisticValueState(op->getContext());
+    knowledge.dtype = Float32Type::get(op->getContext());
+    if (operands[0]->getValue().dtype.isa<Float64Type>()) {
```

Comment:
I think the tests are failing because of this line. I forgot that `operands[0]->getValue().dtype` can be a `nullptr`. You can either check that it is not null before doing `isa<...>()`, or do what you were doing before where you compare to a new `Float64Type`.

Reply:
No problem!

Reply:
I have one more question.
What should the output dtype be if the input dtype is `nullptr` for these ops?
If a `null` input dtype means the input could be any dtype, then I think the output dtype should also be `null` .

Reply:
That's a really good point! Yeah, I agree that `dtype` should indeed be `null` if input is `null`. I think what you can do is to first check if the input `dtype` is null. If so, you can just return the `getNotNonePessimisticValueState` knowledge. Otherwise you can determine which float type to return.